[TEP0100] Remove `EmbeddedStatus` Feature Flag

JeromeJu opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

This commit removes the embedded-status flag following up on the
[Default to Minimal EmbeddedStatus](https://github.com/tektoncd/pipeline/pull/5934) in the v0.44 release. It removes all
the test cases related with the `full` and `both` `EmbeddedStatus` and
only kept the `minimal` test cases as default.

In detail, the test cases that checks the `taskruns` or `runs` status from
previous was changed to checking referenced `taskruns` or `runs` from
the `ChildReferences`.
<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
`embedded-status` feature flag is removed. TaskRun and Run status will be populated only in `pipelineRun.status.childReferences`.
```

JeromeJu commented:
/kind misc

JeromeJu commented:
/ok-to-test

lbernick commented:
> Put up to decide if the ones for the plumbing repo should go in first.

There's no need to change anything in plumbing afaik. Clients should be able to upgrade to the version of pipelines with these changes without taking any additional steps, unless for some reason they rely on `pipelinerun.status.taskruns`.

JeromeJu commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -6607,9 +5860,9 @@ func TestReconcilePipeline_FinalTasks(t *testing.T) {
-				if d := cmp.Diff(reconciledRun.Status.TaskRuns, tt.expectedTaskRuns); d != "" {
-					t.Fatalf("Expected PipelineRunTaskRun status to match TaskRun(s) status, but got a mismatch for %s: %s", tt.name, d)
-				}
+				// if d := cmp.Diff(reconciledRun.Status.TaskRuns, tt.expectedTaskRuns); d != "" {
```
Wondering if it is just safe here to remove the expected TaskRun, not sure if any checks needs to be added functionailty-wise?

lbernick replied:
you could use the tekton client to query the status of the taskruns named in the pipelinerun.status.childreferences and assert that their statuses are as expected.

JeromeJu commented:
For the unit test [`TestReconcilePipeline_FinalTasks`](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L6209), I tried to use the queried taskruns in pipelinerun.status.childreferences to assert but found when we deleted the [updateTaskRunsStatusDirectly](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go#L849), the Conditions will not be updated, which corresponds to the TEP100, since it has been deprecated:
```
This struct, and ChildStatusReferences.ConditionChecks, will be removed once Conditions, which have been deprecated, are removed completely. We are not using child references for the conditions' statuses, because ConditionCheckStatus, the only thing in PipelineRunConditionCheckStatus other than the ConditionName, isn't replicated anywhere else, and contains a fairly minimal amount of data - the pod name, start and completion times, and a corev1.ContainerState. See [the issue for deprecating Conditions](https://github.com/tektoncd/community/blob/main/teps/issue-3377) for more information on the planned removal of Conditions.
```

So I would like to confirm that this is the case and I can just remove the Condition check for `taskruns` status in the first 3 test cases of [`TestReconcilePipeline_FinalTasks`](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L6209)?

lbernick commented:
> For the unit test [`TestReconcilePipeline_FinalTasks`](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L6209), I tried to use the queried taskruns in pipelinerun.status.childreferences to assert but found when we deleted the [updateTaskRunsStatusDirectly](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go#L849), the Conditions will not be updated, which corresponds to the TEP100, since it has been deprecated:
> 
> ```
> This struct, and ChildStatusReferences.ConditionChecks, will be removed once Conditions, which have been deprecated, are removed completely. We are not using child references for the conditions' statuses, because ConditionCheckStatus, the only thing in PipelineRunConditionCheckStatus other than the ConditionName, isn't replicated anywhere else, and contains a fairly minimal amount of data - the pod name, start and completion times, and a corev1.ContainerState. See [the issue for deprecating Conditions](https://github.com/tektoncd/community/blob/main/teps/issue-3377) for more information on the planned removal of Conditions.
> ```
> 
> So I would like to confirm that this is the case and I can just remove the Condition check for `taskruns` status in the first 3 test cases of [`TestReconcilePipeline_FinalTasks`](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L6209)?

Are you sure that the conditions here are the CRD that we removed, rather than the conditions in the taskrun status? There's a bit of name overloading which can be confusing :)

You're correct that the childReferences don't have the status of the child TaskRuns, which is needed in the test. What you can do is get the names of the taskruns from the childReferences, and use a fake tekton client to get these taskruns via a fake api call, asserting that the taskruns returned have the correct status. Here's [an example](https://github.com/tektoncd/pipeline/blob/66cdefa9ce7485a0c39dd9db83d88f2c4f31cba0/pkg/reconciler/pipelinerun/pipelinerun_test.go#L7247) of using the fake tekton client to "fetch" a pipelinerun.

JeromeJu commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -1610,18 +1533,22 @@ status:
+	}
+
+	// Check the status of taskruns from ChildReferences to be Succeeded
+	trFromChildRef, err := clients.Pipeline.TektonV1beta1().TaskRuns("foo").Get(prt.TestAssets.Ctx, reconciledRun.Status.ChildReferences[0].Name, metav1.GetOptions{})
```
This is for comparing with the FinalTask testcase.

JeromeJu commented:
> > For the unit test [`TestReconcilePipeline_FinalTasks`](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L6209), I tried to use the queried taskruns in pipelinerun.status.childreferences to assert but found when we deleted the [updateTaskRunsStatusDirectly](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go#L849), the Conditions will not be updated, which corresponds to the TEP100, since it has been deprecated:
> > ```
> > This struct, and ChildStatusReferences.ConditionChecks, will be removed once Conditions, which have been deprecated, are removed completely. We are not using child references for the conditions' statuses, because ConditionCheckStatus, the only thing in PipelineRunConditionCheckStatus other than the ConditionName, isn't replicated anywhere else, and contains a fairly minimal amount of data - the pod name, start and completion times, and a corev1.ContainerState. See [the issue for deprecating Conditions](https://github.com/tektoncd/community/blob/main/teps/issue-3377) for more information on the planned removal of Conditions.
> > ```
> > 

> > So I would like to confirm that this is the case and I can just remove the Condition check for `taskruns` status in the first 3 test cases of [`TestReconcilePipeline_FinalTasks`](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L6209)?
> 
> Are you sure that the conditions here are the CRD that we removed, rather than the conditions in the taskrun status? There's a bit of name overloading which can be confusing :)
> 
> You're correct that the childReferences don't have the status of the child TaskRuns, which is needed in the test. What you can do is get the names of the taskruns from the childReferences, and use a fake tekton client to get these taskruns via a fake api call, asserting that the taskruns returned have the correct status. Here's [an example](https://github.com/tektoncd/pipeline/blob/66cdefa9ce7485a0c39dd9db83d88f2c4f31cba0/pkg/reconciler/pipelinerun/pipelinerun_test.go#L7247) of using the fake tekton client to "fetch" a pipelinerun.

Thanks @lbernick , I think it is able to compare with the taskRun status referenced as in ChildReference's Name: one changed I made was at https://github.com/tektoncd/pipeline/pull/6049#discussion_r1092064877 if that looks correct as you pointed.

But for the [TestReconcilePipeline_FinalTasks](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L6209), it felt different since when looking in the stacktrace, if it's with test case `pipelineRunStatusFalse`, it hasn't reached https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go#L230 where `ChildReferences` are updated, so we could not check the Status by ChildReference?

JeromeJu commented:
PTAL 👼
/assign @abayer @jerop @lbernick

jerop reviewed (commented):
@JeromeJu I think this change should also remove `TaskRuns` and `Runs` in `PipelineRun.Status`, but it's not clear if that's the case 🤔

lbernick commented on pkg/apis/pipeline/v1beta1/pipelinerun_conversion_test.go:
```diff
@@ -570,14 +507,13 @@ func TestPipelineRunConversionEmbeddedStatusRoundTrip(t *testing.T) {
 }
 
-func TestPipelineRunConversionEmbeddedStatusConvertTo(t *testing.T) {
+func TestPipelineRunConversionConvertTo(t *testing.T) {
```
I think this test could probably be removed

lbernick commented on pkg/reconciler/pipelinerun/cancel.go:
```diff
@@ -192,43 +191,24 @@ func cancelPipelineTaskRunsForTaskNames(ctx context.Context, logger *zap.Sugared
 }
 
 // getChildObjectsFromPRStatusForTaskNames returns taskruns, customruns, and runs in the PipelineRunStatus's ChildReferences or TaskRuns/Runs,
```
```suggestion
// getChildObjectsFromPRStatusForTaskNames returns taskruns, customruns, and runs in the PipelineRunStatus's ChildReferences,
```

JeromeJu commented:
> 

Thanks @jerop.
Sorry I thought we were only removing the feature flag in this release(v0.45). 
Are we targeting to also remove the deprecated fields `status.taskruns` and `status.runs` within the same release? I will open up a separate PR for doing so right now.

lbernick commented on pkg/reconciler/pipelinerun/cancel_test.go:
```diff
@@ -38,196 +38,15 @@ import (
+		runs        []*v1alpha1.Run
+		customRuns  []*v1beta1.CustomRun
+		wantErr     bool
 	}{{
```
I don't think all of these test cases can be deleted; we need to make sure all these cases are still covered. For example, I don't think there's currently a test case for "no resolved taskrun", since "no-resolved-taskrun-with-full" was deleted. Can you take a closer look at these to figure out which are redundant without an embedded-status flag, and which need to be updated to use childreferences?

JeromeJu replied:
restored, sorry that I thought the ones were relied on taskruns status.

JeromeJu commented:
It would be appreciated if we could add this to the [v0.45](https://github.com/tektoncd/pipeline/milestone/64)?

lbernick commented on pkg/reconciler/pipelinerun/cancel_test.go:
```diff
@@ -458,98 +272,7 @@ func TestGetChildObjectsFromPRStatusForTaskNames(t *testing.T) {
-			embeddedStatus: config.MinimalEmbeddedStatus,
+			name: "default",
 			prStatus: v1beta1.PipelineRunStatus{PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
 				TaskRuns: map[string]*v1beta1.PipelineRunTaskRunStatus{
```
No tests should have pipelineruns with taskruns in the status anymore. This test name should also be updated to reflect what it's testing

JeromeJu replied:
Thanks, updated.

lbernick commented on pkg/reconciler/pipelinerun/pipelinerun.go:
```diff
@@ -1369,26 +1318,13 @@ func (c *Reconciler) updatePipelineRunStatusFromInformer(ctx context.Context, pr
-		updatePipelineRunStatusFromChildRefs(logger, pr, taskRuns, runObjects)
-	}
+	// Clear any TaskRuns and Runs present in the status.
+	// This can occur if the value of "embedded-status" was modified during PipelineRun execution prior to its removal.
```
Is this still necessary without an embedded-status flag?

JeromeJu replied:
I was trying to be safe earlier to not to delete the guards before removing the `status.taskruns` or `status.runs` from the apis and also to create the cleanup issue as TODO. I am happy to clean them up if we are doing so in this release.

JeromeJu replied:
I think this could still be valuable before we remove the `status.runs` `status.taskruns` corresponding to the test case we might want at https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_updatestatus_test.go#L1028-L1058?

lbernick replied:
I think we should remove or rewrite test cases with taskruns/runs

lbernick commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -203,40 +202,10 @@ func getPipelineRunUpdates(t *testing.T, actions []ktesting.Action) []*v1beta1.P
 // it checks the resulting API actions, status and events.
-func runTestReconcileWithEmbeddedStatus(t *testing.T, embeddedStatus string) {
+func TestReconcile(t *testing.T) {
 	t.Helper()
```
t.Helper should be removed, here and in all the other test classes that are no longer helpers.

lbernick commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -4943,13 +4661,13 @@ spec:
 	reconciledRun, _ := prt.reconcileRun("foo", "test-pipeline-run-finally-results", []string{}, false)
 
-	expectedPrFullStatus := parse.MustParseV1beta1PipelineRun(t, `
+	expectedPrMinimalStatus := parse.MustParseV1beta1PipelineRun(t, `
```
```suggestion
	expectedPrStatus := parse.MustParseV1beta1PipelineRun(t, `
```

lbernick commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -6607,27 +5918,55 @@ func TestReconcilePipeline_FinalTasks(t *testing.T) {
+		} else if d := cmp.Diff(*expectedTaskRunStatus.Status, trFromChildRef.Status); d != "" {
+			t.Fatalf("Expected TaskRun Status to match ChildReferences TaskRun Status, but got a mismatch %s", diff.PrintWantGot(d))
+		}
+		delete(expectedTaskRuns, trName)
```
Is this necessary? I prefer to avoid modifying inputs unless the function is designed to do so

JeromeJu replied:
Thanks for the suggestion, I delete it so as to makes sure that all expected status are compared.

I changed to using a var for counting if that is better.

lbernick commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -10963,39 +10126,30 @@ spec:
-					t.Errorf("Expected one CustomRun or TaskRun in child references, but found %d", len(pr.Status.ChildReferences))
-				}
-			}
+	if len(pr.Status.Runs) > 1 {
+		t.Errorf("Expected one CustomRun in status, but found %d", len(pr.Status.Runs))
+	}
+	if len(pr.Status.TaskRuns) > 0 {
+		t.Errorf("Expected no TaskRuns in status, but found %d", len(pr.Status.TaskRuns))
+	}
```
This doesn't look right, shouldn't we be checking childreferences?

JeromeJu replied:
Thanks for catching:)

lbernick commented on pkg/reconciler/pipelinerun/timeout_test.go:
```diff
@@ -33,135 +33,14 @@ import (
 		customRuns            []*v1beta1.CustomRun
 		runs                  []*v1alpha1.Run
 		wantErr               bool
 	}{{
```
Same comment as for cancellation tests, I'm not sure all of these test cases can be safely removed

lbernick commented on test/conversion_test.go:
```diff
@@ -1013,14 +986,6 @@ func TestPipelineRunCRDConversion(t *testing.T) {
-
 // validatePipelineRunTaskRunStatusAnnotations validates the `status.taskRun` in Annotations by converting
 // the json string into a struct and compare the valid status fields
 func validatePipelineRunTaskRunStatusAnnotations(pipelineRunName string, annotations map[string]string) error {
```
Is this still needed?

JeromeJu replied:
Thanks for spotting this. I hesitated to remove this since there could be still users using `taskruns` / `runs` for the time being? If so I could remove this in the PR that does the removal?

JeromeJu commented:
I might have misunderstanding on the goal of [v0.45 release for the implementations for TEP0100: ](https://github.com/tektoncd/pipeline/issues/4954): 

- Are we also going to remove `status.taskruns` or `status.runs` within this release v0.45?
- If so, will it be better practice to put it in another PR?

JeromeJu commented:
I also noticed https://github.com/tektoncd/pipeline/pull/4850 for changes in `pkg/status`, I think it might be reasonable to remove those in the following up PR that removes the taskruns and runs fields?

JeromeJu commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -10963,39 +10110,39 @@ spec:
+		case "Run":
+			runsCount++
+		}
+		if runsCount > 1 {
```
TODO: figure out if this is the intended cases or I shall change to case "CustomRun"

JeromeJu replied:
Thanks for checking @XinruZhang

lbernick commented on pkg/reconciler/pipelinerun/cancel_test.go:
```diff
@@ -38,196 +38,131 @@ import (
-	}, {
-		name:           "child-references-with-minimal",
-		embeddedStatus: config.DefaultEmbeddedStatus,
+		name: "child-references",
```
can you rename the test cases with "child-references" in the name to something more descriptive? All of the test cases now have child references, so putting it in the test name isn't very informative.

lbernick commented on pkg/reconciler/pipelinerun/cancel_test.go:
```diff
@@ -458,102 +388,8 @@ func TestGetChildObjectsFromPRStatusForTaskNames(t *testing.T) {
-		}, {
-			name:           "both embedded",
-			embeddedStatus: config.BothEmbeddedStatus,
+			name: "default",
```
can you rename this test to something more descriptive as well?

lbernick commented on pkg/reconciler/pipelinerun/pipelinerun_updatestatus_test.go:
```diff
@@ -935,99 +926,71 @@ pipelineTaskName: task-6
+
+			actualPrStatus := pr.Status
+
+			// The TaskRun keys for recovered taskruns will contain a new random key, appended to the
+			// base name that we expect. Replace the random part so we can diff the whole structure
+			actualTaskRuns := actualPrStatus.PipelineRunStatusFields.TaskRuns
+			if actualTaskRuns != nil {
+				fixedTaskRuns := make(map[string]*v1beta1.PipelineRunTaskRunStatus)
+				re := regexp.MustCompile(`^[a-z\-]*?-task-[0-9]`)
+				for k, v := range actualTaskRuns {
+					newK := re.FindString(k)
+					fixedTaskRuns[newK+"-xxyyy"] = v
 				}
+				actualPrStatus.PipelineRunStatusFields.TaskRuns = fixedTaskRuns
+			}
```
Can this be removed?

JeromeJu replied:
thanks, removed!

lbernick commented on pkg/reconciler/pipelinerun/timeout_test.go:
```diff
@@ -98,70 +112,28 @@ func TestTimeoutPipelineRun(t *testing.T) {
-	}, {
-		name:           "child-references-with-minimal",
-		embeddedStatus: config.MinimalEmbeddedStatus,
+		name: "child-references",
```
same comment here about descriptive test names

lbernick commented on pkg/status/status.go:
```diff
@@ -77,9 +77,9 @@ func GetRunStatusForPipelineTask(ctx context.Context, client versioned.Interface
+// GetPipelineTaskStatuses returns populated TaskRun and Run status maps for a PipelineRun from its ChildReferences.
 // If the PipelineRun has no ChildReferences, its .Status.TaskRuns and .Status.Runs will be returned instead.
-func GetFullPipelineTaskStatuses(ctx context.Context, client versioned.Interface, ns string, pr *v1beta1.PipelineRun) (map[string]*v1beta1.PipelineRunTaskRunStatus,
+func GetPipelineTaskStatuses(ctx context.Context, client versioned.Interface, ns string, pr *v1beta1.PipelineRun) (map[string]*v1beta1.PipelineRunTaskRunStatus,
```
@abayer is this meant to be a helper function for package importers outside tektoncd/pipeline? It seems like dead code.

@JeromeJu can you please update this function and its docstring to not use taskruns/runs?

JeromeJu replied:
I was not sure if this is used for populating taskruns/runs as `full` embbededStatus alike pipelineRunStatus to prepare for the removal of this feature flag? Was also wondering about this at
https://github.com/tektoncd/pipeline/issues/6090#issuecomment-1412730427

JeromeJu replied:
Updated and removed the populating from taskruns/runs. Thanks for the pointers

lbernick commented on pkg/status/status_test.go:
```diff
@@ -295,7 +295,7 @@ status:
 		},
 		{
-			name: "minimal embedded",
+			name: "default",
```
can you rename this to something more descriptive?

JeromeJu replied:
Thanks, renamed.

lbernick commented:
@JeromeJu just edited the release notes since we don't really have a way for people to migrate off of the feature flag; they just can't use it anymore

JeromeJu commented:
> @JeromeJu just edited the release notes since we don't really have a way for people to migrate off of the feature flag; they just can't use it anymore

Ohhh thanks for correcting that:)

chitrangpatel reviewed (approved):
Looks good! Just wanted to ask what happens if users provide `embedded-status` flag. Do they get an error or does it just silently ignore it?

JeromeJu commented:
> Looks good! Just wanted to ask what happens if users provide `embedded-status` flag. Do they get an error or does it just silently ignore it?

I don't think they will get an error as it shall be just an unknown feature flag in config/config-feature-flag

chitrangpatel commented:
> > Looks good! Just wanted to ask what happens if users provide `embedded-status` flag. Do they get an error or does it just silently ignore it?
> 
> I don't think they will get an error as it shall be just an unknown feature flag in config/config-feature-flag

I think thats ok.

jerop reviewed (commented):
@JeromeJu please also need to remove `embedded-status` feature flag from dogfooding - https://github.com/tektoncd/plumbing/blob/main/tekton/cd/pipeline/overlays/dogfooding/feature-flags.yaml#L8

thank you!

/lgtm

JeromeJu commented:
/test pull-tekton-pipeline-beta-integration-tests

JeromeJu commented:
/retest
to retrigger ci for build test

JeromeJu commented:
/retest

JeromeJu commented:
Thanks Jerop, raised https://github.com/tektoncd/plumbing/pull/1339 for followup at plumbing repo.
